src/AbstractActionHandler.ts

```diff
@@ -192,19 +204,35 @@ export abstract class AbstractActionHandler {
     block: Block,
     context: any,
     isReplay: boolean,
+    lastIrreversibleBlockNumber: number,
   ): Promise<void> {
     const { blockInfo } = block
 
-    const versionedActions = await this.applyUpdaters(state, block, isReplay, context)
+    const versionedActions = await this.applyUpdaters(state, block, context, isReplay)
     if (!isReplay) {
-      this.runEffects(versionedActions, block, context)
+      this.runEffects(versionedActions, block, context, lastIrreversibleBlockNumber)
     }
 
     await this.updateIndexState(state, block, isReplay, this.handlerVersionName, context)
     this.lastProcessedBlockNumber = blockInfo.blockNumber
     this.lastProcessedBlockHash = blockInfo.blockHash
   }
 
+  private range(start: number, end: number) {
+    return Array(end - start).fill(0).map((_, i: number) => i + start)
+  }
+
+  private runDeferredEffects(lastIrreversibleBlockNumber) {
+    for (const blockNumber of this.range(this.nextDeferredBlockNumber, lastIrreversibleBlockNumber + 1)) {
+      if (this.deferredEffects[blockNumber]) {
+        for (const deferredEffect of this.deferredEffects[blockNumber]) {
+          deferredEffect()
```

Comment:
does a deferred effect need to be executed sequentially? or can it be in parallel, if we're not awaiting completion it could end up race conditioning.  I've been out of demux for a while so I'm unsure as to when order is and isn't important.

Reply:
Since we're not awaiting, this is equivalent. It would just be a code style change.

Reply:
Order shouldn't be required for effects